File: bindings/kotlin/InterShareSDK/src/main/java/com/julian_baumann/intershare_sdk/Discovery.kt
package com.julian_baumann.intershare_sdk

import android.content.Context
import com.julian_baumann.intershare_sdk.bluetoothLowEnergy.BLECentralManager

interface DiscoveryDelegate: DeviceListUpdateDelegate

class Discovery(context: Context, delegate: DiscoveryDelegate) {
    private val internal: InternalDiscovery = InternalDiscovery(delegate)
    private val bleImplementation: BLECentralManager = BLECentralManager(context, internal)

    init {
        internal.addBleImplementation(bleImplementation)
    }

    fun getDevices(): List<Device> {
        return internal.getDevices()
    }

    fun startScanning() {
        internal.start()
    }

    fun stopScanning() {
        internal.stop()
    }
}

File: src/intershare_windows/src/discovery.rs
use crate::ble::ble_client::BleClient;
pub use intershare_sdk::discovery::Discovery as InternalDiscovery;
use intershare_sdk::{init_logger, Device, DiscoveryDelegate};
use std::sync::{Arc, Mutex};
use intershare_sdk::errors::DiscoverySetupError;

pub struct Discovery {
    internal_discovery: Arc<Mutex<InternalDiscovery>>,
}

impl Discovery {
    pub fn new(delegate: Option<Box<dyn DiscoveryDelegate>>) -> Result<Self, DiscoverySetupError> {
        init_logger();
        let internal_discovery = Arc::new(Mutex::new(
            InternalDiscovery::new(delegate)?,
        ));

        let ble_implementation = BleClient::new(internal_discovery.clone());

        {
            let mut internal_discovery_mut = internal_discovery.lock().unwrap();
            internal_discovery_mut.add_ble_implementation(Box::new(ble_implementation));
        }

        Ok(Self {
            internal_discovery,
        })
    }

    pub fn get_devices(&self) -> Vec<Device> {
        let internal_discovery = self.internal_discovery.lock().unwrap();
        internal_discovery.get_devices()
    }

    pub fn start(&self) {
        let internal_discovery = self.internal_discovery.lock().unwrap();
        internal_discovery.start()
    }

    pub fn stop(&self) {
        let internal_discovery = self.internal_discovery.lock().unwrap();
        internal_discovery.stop()
    }
}
